### ISSUE TITLE
Before(Assembly) hook does not wait for Before(TestSession) hook

### ISSUE THREAD

**Issue by connorivy** [thomhurst/TUnit#1444]
Not sure if this is an issue or by design.

I have two hooks that I want to run to setup my api before I start testing

```
[Before(TestSession)]
public async Task SetupApi()
{
    // async code
}

[Before(Assembly)]
public async Task SeedDb()
{
    // async code
}
```

These hooks are called in the correct order, first the BeforeTestSession and then the BeforeAssembly, but the BeforeAssembly setup is called before the BeforeTestSession finishes. 

Is there a way to chain hooks so that each new link of the chain fires after the previous has completed?

**Comment by thomhurst** [comment]
Would you be able to create a minimal repro please

**Comment by thomhurst** [comment]
I can't replicate. 

Assembly 1:

```csharp
public class Tests : BaseClass
{
    [Before(TestDiscovery)]
    public static void EnsureLibAssemblyLoaded()
    {
        Console.WriteLine(typeof(BaseClass).Assembly.FullName);
    }
    
    [Before(TestSession)]
    public static async Task BeforeTestSession()
    {
        await StringWriter.WriteLineAsync($"BeforeTestSession started at {DateTime.Now}");
        await Task.Delay(TimeSpan.FromSeconds(5));
        await StringWriter.WriteLineAsync($"BeforeTestSession ended at {DateTime.Now}");
    }

    [Before(Assembly)]
    public static async Task BeforeAssembly()
    {
        await StringWriter.WriteLineAsync($"BeforeAssembly_Assembly1 started at {DateTime.Now}");
        await Task.Delay(TimeSpan.FromSeconds(5));
        await StringWriter.WriteLineAsync($"BeforeAssembly_Assembly1 ended at {DateTime.Now}");
    }

    [Test]
    public void Test1()
    {
        StringWriter.WriteLine($"Executing Test_Assembly1 at {DateTime.Now}");
    }
}
```

Assembly 2:

```csharp
public class BaseClass
{
    protected static readonly StringWriter StringWriter = new();
}

public class Tests : BaseClass
{
    [Before(Assembly)]
    public static async Task BeforeAssembly()
    {
        await StringWriter.WriteLineAsync($"BeforeAssembly_Assembly2 started at {DateTime.Now}");
        await Task.Delay(TimeSpan.FromSeconds(5));
        await StringWriter.WriteLineAsync($"BeforeAssembly_Assembly2 ended at {DateTime.Now}");
    }
    
    [Test]
    public void Test1()
    {
        StringWriter.WriteLine($"Executing Test_Assembly2 at {DateTime.Now}");
    }
}
```

Output:

> BeforeTestSession started at 21/12/2024 15:36:01
BeforeTestSession ended at 21/12/2024 15:36:06
BeforeAssembly_Assembly2 started at 21/12/2024 15:36:06
BeforeAssembly_Assembly1 started at 21/12/2024 15:36:06
BeforeAssembly_Assembly1 ended at 21/12/2024 15:36:11
BeforeAssembly_Assembly2 ended at 21/12/2024 15:36:11
Executing Test_Assembly2 at 21/12/2024 15:36:11
Executing Test_Assembly1 at 21/12/2024 15:36:11

**Comment by thomhurst** [comment]
@connorivy ?

**Comment by connorivy** [comment]
Sorry for the delay. I am still experiencing this behavior on a decently complicated project of mine. However, I've tried for a while and I can't seem to create a minimal reproducible example. When I try to dumb it down to a simpler example, your library works exactly as expected.

I think the issue may have something to do with the fact that one of the two test projects references the other, and I am usually running both of the test projects at the same time. But I did manage to "fix" the issue using the "EnsureLibAssemblyLoaded" method from your above example. I didn't know you could do that. Now instead of running both test assemblies that have a reference relationship, I am just running the referencing assembly tests which now contain the tests of the referenced assemblies.

Thank you for responding and I apologize for this issue being a bit of a wild goose chase

**Comment by thomhurst** [comment]
Yeah that'll be the issue, you were running two test suites so they were separate from each other.

Assemblies are loaded into the memory when first referenced, so by just referencing your assembly by calling a Type etc, it'll load it in. Just need to make sure you do it before test discovery so it finds those tests for ya 😁
